File: 알고리즘/무개과제/002. ㅋㅋㅋ퀵솔트/ㅋㅋㅋ퀵솔트.py
def partition(arr, start, end):
    pivot = arr[end]
    left = start
    right = end - 1

    while True:
        while left <= right and arr[left] < pivot:
            left += 1
        while left <= right and arr[right] > pivot:
            right -= 1
        if left < right:
            arr[left], arr[right] = arr[right], arr[left]
        else:
            break
    arr[end], arr[left] = arr[left], arr[end]
    return left


def quick_sort(arr, start, end):
    if start < end and not is_sorted(arr, start, end):
        pivot = partition(arr, start, end)
        quick_sort(arr, start, pivot - 1)
        quick_sort(arr, pivot + 1, end)


def is_sorted(arr, start, end):
    for i in range(start, end):
        if arr[i] > arr[i + 1]:
            return False
    return True

File: 알고리즘/무개과제/002. ㅋㅋㅋ퀵솔트/ㅋㅋㅋ퀵솔트.kt
fun quickSort(arr: IntArray, left: Int, right: Int) {
    if (left >= right || isSorted(arr, left, right)) {
        return
    }

    val middle = (left + right) / 2
    val pivot = arr[middle]
    var l = left
    var r = right

    while (l <= r) {
        while (arr[l] < pivot) {
            l++
        }
        while (arr[r] > pivot) {
            r--
        }
        if (l <= r) {
            arr[l] = arr[r].also { arr[r] = arr[l] }
            l++
            r--
        }
    }
    if (left < r && !isSorted(arr, left, r)) {
        quickSort(arr, left, r)
    }
    if (right > l && !isSorted(arr, l, right)) {
        quickSort(arr, l, right)
    }
}

// 중간에 배열이 다 정렬되면 정렬 그만하기 위한 조건 검사 함수
fun isSorted(arr: IntArray, left: Int, right: Int): Boolean {
    for (i in left..<right) {
        if (arr[i] > arr [i + 1]) {
            return false
        }
    }
    return true
}
